File: windows/ProfileStore.cs
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Windows.Media;
using System.Windows.Threading;

namespace CursorProfiles;

/// <summary>Wraps a profile with live UI state (size, running, cloning).</summary>
public class ProfileVM : INotifyPropertyChanged
{
    public CursorProfile Model { get; private set; }

    public ProfileVM(CursorProfile model) => Model = model;

    private long? _sizeBytes;
    private List<int> _runningPids = new();
    private bool _isDuplicating;

    public event PropertyChangedEventHandler? PropertyChanged;
    private void Raise([CallerMemberName] string? name = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));

    public void Update(CursorProfile model)
    {
        Model = model;
        RaiseAll();
    }

    public void RaiseAll()
    {
        foreach (var p in new[] { nameof(DisplayName), nameof(Emoji), nameof(HeaderBrush),
                 nameof(AccentBrush), nameof(SizeText), nameof(LastLaunchedText), nameof(IsRunning),
                 nameof(IsSystem), nameof(ShowPin), nameof(ProjectText), nameof(HasProject),
                 nameof(RunningText), nameof(DotBrush), nameof(LaunchLabel), nameof(IsDuplicating) })
            Raise(p);
    }

    public long? SizeBytes
    {
        get => _sizeBytes;
        set { _sizeBytes = value; Raise(nameof(SizeText)); }
    }

    public List<int> RunningPids
    {
        get => _runningPids;
        set
        {
            _runningPids = value;
            Raise(nameof(IsRunning)); Raise(nameof(RunningText));
            Raise(nameof(DotBrush)); Raise(nameof(LaunchLabel));
        }
    }

    public bool IsDuplicating
    {
        get => _isDuplicating;
        set { _isDuplicating = value; Raise(); }
    }

    // Bindable projections
    public string DisplayName => Model.DisplayName;
    public string Emoji => Model.Emoji;
    public bool IsSystem => Model.IsSystem;
    public bool ShowPin => Model.IsPinned && !Model.IsSystem;
    public bool IsRunning => _runningPids.Count > 0;
    public string RunningText => IsRunning ? "Running" : "Idle";
    public string LaunchLabel => IsRunning ? "New Window" : "Launch";
    public string SizeText => Format.Bytes(_sizeBytes);
    public string LastLaunchedText => Format.Relative(Model.LastLaunchedAt);
    public string ProjectText => Model.DefaultProjectPath ?? "";
    public bool HasProject => !string.IsNullOrEmpty(Model.DefaultProjectPath);

    public Brush AccentBrush => new SolidColorBrush(Palette.ColorFromHex(Model.ColorHex));
    public Brush DotBrush => IsRunning
        ? new SolidColorBrush(Color.FromRgb(0x4A, 0xDE, 0x80))
        : new SolidColorBrush(Color.FromArgb(0x80, 0xFF, 0xFF, 0xFF));

    public Brush HeaderBrush
    {
        get
        {
            var c = Palette.ColorFromHex(Model.ColorHex);
            var dim = Color.FromArgb(0xFF,
                (byte)(c.R * 0.72), (byte)(c.G * 0.72), (byte)(c.B * 0.72));
            return new LinearGradientBrush(c, dim, 35);
        }
    }
}

public class ProfileStore
{
    public string ProfilesDir { get; }
    public LauncherSettings Settings { get; }
    public List<ProfileVM> Profiles { get; } = new();

    public event Action? Changed;
    public event Action<string>? Error;
    public event Action<string>? Info;

    private const string MetadataName = ".profiles.json";
    private readonly DispatcherTimer _pollTimer;

    public static string SystemDataDir =>
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Cursor");

    public string? ResolvedCursorPath => CursorLauncher.FindCursor(
        string.IsNullOrWhiteSpace(Settings.CustomCursorPath) ? null : Settings.CustomCursorPath);

    public ProfileStore()
    {
        ProfilesDir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".cursor_profiles");
        Settings = LauncherSettings.Load(ProfilesDir);
        Reload();

        _pollTimer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(4) };
        _pollTimer.Tick += (_, _) => RefreshRunning();
        _pollTimer.Start();
    }

    public string DirectoryFor(CursorProfile p) =>
        p.IsSystem ? SystemDataDir : Path.Combine(ProfilesDir, p.FolderName);

    // MARK: Loading

    public void Reload()
    {
        Directory.CreateDirectory(ProfilesDir);

        var known = LoadMetadata();

        var onDisk = Directory.EnumerateDirectories(ProfilesDir)
            .Select(Path.GetFileName)
            .Where(n => n is not null && !n.StartsWith('.'))
            .Cast<string>()
            .ToList();

        foreach (var name in onDisk)
        {
            if (known.Any(p => p.FolderName == name)) continue;
            var display = name.StartsWith("custom_") ? name["custom_".Length..] : name;
            known.Add(new CursorProfile
            {
                FolderName = name,
                DisplayName = display,
                ColorHex = Palette.RandomHex(),
            });
        }
        // Drop metadata for folders that no longer exist (never the built-in one).
        known.RemoveAll(p => !p.IsSystem && !onDisk.Contains(p.FolderName));

        // Surface the user's original Cursor profile — visible, launchable, protected.
        if (!known.Any(p => p.IsSystem) && Directory.Exists(SystemDataDir))
        {
            known.Insert(0, new CursorProfile
            {
                FolderName = CursorProfile.SystemFolderName,
                DisplayName = "Main Cursor",
                Emoji = "⭐",
                ColorHex = "#3B82F6",
                IsPinned = true,
                IsSystem = true,
            });
        }

        Profiles.Clear();
        foreach (var p in known) Profiles.Add(new ProfileVM(p));
        SaveMetadata();
        Changed?.Invoke();
        RefreshSizes();
        RefreshRunning();
    }

    private List<CursorProfile> LoadMetadata()
    {
        try
        {
            var path = Path.Combine(ProfilesDir, MetadataName);
            if (File.Exists(path))
                return JsonSerializer.Deserialize<List<CursorProfile>>(File.ReadAllText(path)) ?? new();
        }
        catch { /* corrupted metadata — rebuild from folders */ }
        return new();
    }

    public void SaveMetadata()
    {
        try
        {
            var json = JsonSerializer.Serialize(
                Profiles.Select(p => p.Model).ToList(),
                new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(Path.Combine(ProfilesDir, MetadataName), json);
        }
        catch (Exception e) { Error?.Invoke($"Could not save profile metadata: {e.Message}"); }
    }

    // MARK: CRUD

    public ProfileVM? CreateProfile(string displayName, string emoji, string colorHex,
                                    int memoryMB, string? projectPath)
    {
        var baseName = ProfileNaming.SanitizeFolderName(displayName.Replace(' ', '_'));
        if (baseName.Length == 0)
        {
            Error?.Invoke("Profile name must contain at least one letter, number, hyphen or underscore.");
            return null;
        }
        var folder = UniqueFolderName(baseName);

        var profile = new CursorProfile
        {
            FolderName = folder,
            DisplayName = displayName.Trim(),
            Emoji = emoji,
            ColorHex = colorHex,
            DefaultMemoryMB = memoryMB,
            DefaultProjectPath = string.IsNullOrWhiteSpace(projectPath) ? null : projectPath,
        };
        try { Directory.CreateDirectory(DirectoryFor(profile)); }
        catch (Exception e)
        {
            Error?.Invoke($"Could not create profile folder: {e.Message}");
            return null;
        }
        var vm = new ProfileVM(profile);
        Profiles.Add(vm);
        SaveMetadata();
        Changed?.Invoke();
        RefreshSizes();
        TitleBarColorizer.Apply(profile, DirectoryFor(profile));
        return vm;
    }

    private string UniqueFolderName(string baseName)
    {
        var folder = baseName;
        var counter = 2;
        while (folder == CursorProfile.SystemFolderName ||
               Profiles.Any(p => string.Equals(p.Model.FolderName, folder, StringComparison.OrdinalIgnoreCase)))
        {
            folder = $"{baseName}-{counter++}";
        }
        return folder;
    }

    public void Update(ProfileVM vm, CursorProfile updated)
    {
        var before = vm.Model;
        var colorChanged = before.ColorHex != updated.ColorHex;
        var cosmeticsChanged = before.DisplayName != updated.DisplayName
            || before.Emoji != updated.Emoji
            || colorChanged
            || before.DefaultMemoryMB != updated.DefaultMemoryMB
            || before.DefaultProjectPath != updated.DefaultProjectPath;
        vm.Update(updated);
        SaveMetadata();
        Changed?.Invoke();
        // Keep an existing shortcut in sync with name/icon/launch changes.
        if (cosmeticsChanged && !updated.IsSystem && HasShortcut(vm))
            CreateOrUpdateShortcut(vm, announce: false);
        if (colorChanged)
            TitleBarColorizer.Apply(updated, DirectoryFor(updated));
    }

    // MARK: Taskbar shortcuts

    public bool HasShortcut(ProfileVM vm) => ShortcutBuilder.Exists(vm.Model);

    /// <summary>Build (or rebuild) a Start Menu shortcut with a generated icon
    /// for this profile, so it can be pinned to the taskbar.</summary>
    public void CreateOrUpdateShortcut(ProfileVM vm, bool announce = true)
    {
        try
        {
            ShortcutBuilder.CreateOrUpdate(vm.Model, DirectoryFor(vm.Model), Settings.CustomCursorPath);
            if (announce)
            {
                Info?.Invoke(
                    $"A shortcut for \"{vm.Model.DisplayName}\" was added to the Start Menu " +
                    "(Prism folder). Find it in Start, then right-click it and choose " +
                    "\"Pin to taskbar\".");
            }
        }
        catch (Exception e)
        {
            Error?.Invoke($"Could not create the shortcut: {e.Message}");
        }
    }

    public void RemoveShortcut(ProfileVM vm) => ShortcutBuilder.Remove(vm.Model);

    /// <summary>Duplicate a profile; for the built-in one this clones the original
    /// Cursor data into a new managed profile (the original is only read).</summary>
    public async void Duplicate(ProfileVM vm)
    {
        var profile = vm.Model;
        var baseName = profile.IsSystem ? "Main-Cursor-Clone" : profile.FolderName + "-copy";
        var folder = UniqueFolderName(baseName);
        var source = DirectoryFor(profile);
        var dest = Path.Combine(ProfilesDir, folder);

        var copy = profile.Copy();
        copy.FolderName = folder;
        copy.DisplayName = profile.IsSystem ? "Main Cursor Clone" : profile.DisplayName + " Copy";
        copy.CreatedAt = DateTime.UtcNow;
        copy.LastLaunchedAt = null;
        copy.IsPinned = false;
        copy.IsSystem = false;

        vm.IsDuplicating = true;
        try
        {
            await Task.Run(() => CopyDirectory(source, dest));
            Profiles.Add(new ProfileVM(copy));
            SaveMetadata();
            Changed?.Invoke();
            RefreshSizes();
        }
        catch (Exception e)
        {
            Error?.Invoke($"Could not duplicate profile: {e.Message}");
            try { if (Directory.Exists(dest)) Directory.Delete(dest, true); } catch { }
        }
        finally { vm.IsDuplicating = false; }
    }

    private static void CopyDirectory(string source, string dest)
    {
        Directory.CreateDirectory(dest);
        foreach (var dir in Directory.EnumerateDirectories(source, "*", SearchOption.AllDirectories))
            Directory.CreateDirectory(dir.Replace(source, dest));
        foreach (var file in Directory.EnumerateFiles(source, "*", SearchOption.AllDirectories))
        {
            try { File.Copy(file, file.Replace(source, dest), overwrite: true); }
            catch (IOException) { /* locked cache file — skip */ }
        }
    }

    /// <summary>Send the profile folder to the Recycle Bin (recoverable).</summary>
    public void Delete(ProfileVM vm)
    {
        if (vm.Model.IsSystem)
        {
            Error?.Invoke("The built-in Cursor profile can't be deleted from here — it's your original Cursor data.");
            return;
        }
        if (vm.IsRunning)
        {
            Error?.Invoke($"'{vm.Model.DisplayName}' is running. Quit it before deleting.");
            return;
        }
        try
        {
            Microsoft.VisualBasic.FileIO.FileSystem.DeleteDirectory(
                DirectoryFor(vm.Model),
                Microsoft.VisualBasic.FileIO.UIOption.OnlyErrorDialogs,
                Microsoft.VisualBasic.FileIO.RecycleOption.SendToRecycleBin);
        }
        catch (Exception e)
        {
            Error?.Invoke($"Could not move profile to Recycle Bin: {e.Message}");
            return;
        }
        RemoveShortcut(vm);
        Profiles.Remove(vm);
        SaveMetadata();
        Changed?.Invoke();
    }

    public void TogglePin(ProfileVM vm)
    {
        var m = vm.Model.Copy();
        m.IsPinned = !m.IsPinned;
        Update(vm, m);
    }

    // MARK: Launch / quit

    public void Launch(ProfileVM vm, string? projectPath = null, int? memoryMB = null, bool newWindow = false)
    {
        var p = vm.Model;
        // Idempotent — also covers profiles adopted from disk that never went through CreateProfile.
        TitleBarColorizer.Apply(p, DirectoryFor(p));
        try
        {
            CursorLauncher.Launch(
                userDataDir: p.IsSystem ? null : DirectoryFor(p),
                cursorPath: string.IsNullOrWhiteSpace(Settings.CustomCursorPath) ? null : Settings.CustomCursorPath,
                memoryMB: memoryMB ?? p.DefaultMemoryMB,
                projectPath: projectPath ?? p.DefaultProjectPath,
                newWindow: newWindow);
        }
        catch (Exception e)
        {
            Error?.Invoke(e.Message);
            return;
        }
        var m = p.Copy();
        m.LastLaunchedAt = DateTime.UtcNow;
        Update(vm, m);
        Task.Delay(2000).ContinueWith(_ => RefreshRunning(),
            TaskScheduler.FromCurrentSynchronizationContext());
    }

    public void Quit(ProfileVM vm)
    {
        CursorLauncher.Quit(vm.RunningPids);
        Task.Delay(1500).ContinueWith(_ => RefreshRunning(),
            TaskScheduler.FromCurrentSynchronizationContext());
    }

    // MARK: Background refresh

    public async void RefreshRunning()
    {
        var snapshot = Profiles.Select(vm => (vm, dir: DirectoryFor(vm.Model), vm.Model.IsSystem)).ToList();
        var processes = await Task.Run(CursorLauncher.ProcessList);
        foreach (var (vm, dir, isSystem) in snapshot)
        {
            vm.RunningPids = isSystem
                ? CursorLauncher.SystemProfilePids(processes)
                : CursorLauncher.Pids(processes, dir);
        }
    }

    public async void RefreshSizes()
    {
        var snapshot = Profiles.Select(vm => (vm, dir: DirectoryFor(vm.Model))).ToList();
        foreach (var (vm, dir) in snapshot)
        {
            var size = await Task.Run(() => DirectorySize(dir));
            vm.SizeBytes = size;
        }
    }

    private static long DirectorySize(string path)
    {
        long total = 0;
        try
        {
            foreach (var file in Directory.EnumerateFiles(path, "*", new EnumerationOptions
                     { RecurseSubdirectories = true, IgnoreInaccessible = true }))
            {
                try { total += new FileInfo(file).Length; } catch { }
            }
        }
        catch { }
        return total;
    }
}

File: Sources/CursorProfiles/ProfileStore.swift
import AppKit
import Combine
import SwiftUI

// MARK: - Store

@MainActor
final class ProfileStore: ObservableObject {

    @Published var profiles: [CursorProfile] = []
    @Published var sizes: [String: Int64] = [:]
    @Published var runningPIDs: [String: [Int32]] = [:]
    /// Live resident memory per profile (bytes, summed RSS of its processes).
    @Published var liveMemory: [String: UInt64] = [:]
    /// Live CPU % per profile (summed %cpu of its processes).
    @Published var liveCPU: [String: Double] = [:]
    /// Real on-screen window counts per profile.
    @Published var windowCounts: [String: Int] = [:]
    /// Real system-wide CPU %, nil until the second sample.
    @Published var systemCPU: Double?
    @Published var duplicating: Set<String> = []
    @Published var lastError: String?

    private let cpuMonitor = SystemCPUMonitor()

    /// Real total RAM of this Mac.
    var physicalMemory: UInt64 { SystemStats.physicalMemory }

    /// Real summed RSS of every running profile.
    var liveMemoryTotal: UInt64 { liveMemory.values.reduce(0, +) }

    @AppStorage("customCursorPath") var customCursorPath: String = ""
    @AppStorage("defaultMemoryMB") var defaultMemoryMB: Int = 16384

    let profilesDir: URL
    private let metadataName = ".profiles.json"
    private var pollTask: Task<Void, Never>?

    var resolvedCursorPath: String? {
        CursorLauncher.findCursor(customPath: customCursorPath.isEmpty ? nil : customCursorPath)
    }

    init(profilesDir: URL = URL(fileURLWithPath: NSHomeDirectory()).appendingPathComponent(".cursor_profiles")) {
        self.profilesDir = profilesDir
        reload()
        startPolling()
    }

    /// The built-in Cursor data directory — where Cursor keeps everything when
    /// launched normally. This is the profile people already have.
    static var systemDataDir: URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Cursor")
    }

    func directory(for profile: CursorProfile) -> URL {
        profile.isSystem ? Self.systemDataDir : profilesDir.appendingPathComponent(profile.folderName)
    }

    func isRunning(_ profile: CursorProfile) -> Bool {
        !(runningPIDs[profile.folderName] ?? []).isEmpty
    }

    // MARK: Loading & persistence

    func reload() {
        try? FileManager.default.createDirectory(at: profilesDir, withIntermediateDirectories: true)

        var known = loadMetadata()

        // Adopt any directories created by hand.
        let onDisk = (try? FileManager.default.contentsOfDirectory(
            at: profilesDir, includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles])) ?? []
        let dirNames = onDisk
            .filter { (try? $0.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true }
            .map(\.lastPathComponent)

        for name in dirNames where !known.contains(where: { $0.folderName == name }) {
            var display = name
            if display.hasPrefix("custom_") { display.removeFirst("custom_".count) }
            known.append(CursorProfile(folderName: name, displayName: display))
        }
        // Drop metadata for folders that no longer exist (never the built-in one).
        known.removeAll { !$0.isSystem && !dirNames.contains($0.folderName) }

        // Surface the user's original Cursor profile so it's visible and
        // launchable — and clearly protected — rather than silently ignored.
        if !known.contains(where: { $0.isSystem }),
           FileManager.default.fileExists(atPath: Self.systemDataDir.path) {
            known.insert(CursorProfile(
                folderName: CursorProfile.systemFolderName,
                displayName: "Main Cursor",
                colorHex: "#3B82F6",
                isPinned: true,
                isSystem: true
            ), at: 0)
        }

        profiles = known
        saveMetadata()
        refreshSizes()
        refreshRunning()
    }

    private func loadMetadata() -> [CursorProfile] {
        let url = profilesDir.appendingPathComponent(metadataName)
        guard let data = try? Data(contentsOf: url) else { return [] }
        return (try? JSONDecoder().decode([CursorProfile].self, from: data)) ?? []
    }

    private func saveMetadata() {
        let url = profilesDir.appendingPathComponent(metadataName)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        if let data = try? encoder.encode(profiles) {
            try? data.write(to: url, options: .atomic)
        }
    }

    // MARK: CRUD

    @discardableResult
    func createProfile(displayName: String,
                       colorHex: String,
                       memoryMB: Int,
                       defaultProjectPath: String?) -> CursorProfile? {
        let base = ProfileNaming.sanitizeFolderName(displayName.replacingOccurrences(of: " ", with: "_"))
        guard !base.isEmpty else {
            lastError = "Profile name must contain at least one letter, number, hyphen or underscore."
            return nil
        }
        var folder = base
        var counter = 2
        while profiles.contains(where: { $0.folderName.caseInsensitiveCompare(folder) == .orderedSame })
            || folder == CursorProfile.systemFolderName {
            folder = "\(base)-\(counter)"
            counter += 1
        }

        let profile = CursorProfile(
            folderName: folder,
            displayName: displayName.trimmingCharacters(in: .whitespaces),
            colorHex: colorHex,
            defaultMemoryMB: memoryMB,
            defaultProjectPath: defaultProjectPath
        )
        do {
            try FileManager.default.createDirectory(at: directory(for: profile), withIntermediateDirectories: true)
        } catch {
            lastError = "Could not create profile folder: \(error.localizedDescription)"
            return nil
        }
        profiles.append(profile)
        saveMetadata()
        refreshSizes()
        TitleBarColorizer.apply(profile: profile, profileDir: directory(for: profile))
        return profile
    }

    func update(_ profile: CursorProfile) {
        guard let idx = profiles.firstIndex(where: { $0.id == profile.id }) else { return }
        let colorChanged = profiles[idx].colorHex != profile.colorHex
        profiles[idx] = profile
        saveMetadata()
        if colorChanged {
            TitleBarColorizer.apply(profile: profile, profileDir: directory(for: profile))
        }
    }

    /// Duplicate a profile. For the built-in profile this *clones* the original
    /// Cursor data into a new managed profile — the original is only read,
    /// never touched. Runs in the background since data dirs can be large.
    func duplicate(_ profile: CursorProfile) {
        let base = profile.isSystem
            ? "Main-Cursor-Clone"
            : profile.folderName + "-copy"
        var folder = base
        var counter = 2
        while profiles.contains(where: { $0.folderName.caseInsensitiveCompare(folder) == .orderedSame }) {
            folder = "\(base)-\(counter)"
            counter += 1
        }
        let source = directory(for: profile)
        let dest = profilesDir.appendingPathComponent(folder)

        var copy = profile
        copy.folderName = folder
        copy.displayName = profile.isSystem ? "Main Cursor Clone" : profile.displayName + " Copy"
        copy.createdAt = Date()
        copy.lastLaunchedAt = nil
        copy.isPinned = false
        copy.isSystem = false

        duplicating.insert(profile.folderName)
        // Sendable snapshots cross the thread hop; the finish step runs in
        // this Task, which inherits MainActor isolation from duplicate().
        let folderName = profile.folderName
        let sourceURL = source
        let destURL = dest
        let finishedCopy = copy
        Task { [weak self] in
            let copyErrorMessage = await Task.detached(priority: .userInitiated) {
                Self.copyProfileData(source: sourceURL, dest: destURL)
            }.value
            self?.finishDuplicate(folderName: folderName, copy: finishedCopy,
                                  dest: destURL, errorMessage: copyErrorMessage)
        }
    }

    /// Pure file copy: Sendable in/out, never touches the store.
    private static nonisolated func copyProfileData(source: URL, dest: URL) -> String? {
        do {
            try FileManager.default.copyItem(at: source, to: dest)
            return nil
        } catch {
            return error.localizedDescription
        }
    }

    @MainActor
    private func finishDuplicate(folderName: String, copy: CursorProfile,
                                 dest: URL, errorMessage: String?) {
        duplicating.remove(folderName)
        if let errorMessage {
            lastError = "Could not duplicate profile: \(errorMessage)"
            try? FileManager.default.removeItem(at: dest)
        } else {
            profiles.append(copy)
            saveMetadata()
            refreshSizes()
        }
    }

    /// Moves the profile folder to the Trash (recoverable, unlike the script's rm -rf).
    func delete(_ profile: CursorProfile) {
        if profile.isSystem {
            lastError = "The built-in Cursor profile can't be deleted from here — it's your original Cursor data."
            return
        }
        if isRunning(profile) {
            lastError = "'\(profile.displayName)' is running. Quit it before deleting."
            return
        }
        do {
            try FileManager.default.trashItem(at: directory(for: profile), resultingItemURL: nil)
        } catch {
            lastError = "Could not move profile to Trash: \(error.localizedDescription)"
            return
        }
        profiles.removeAll { $0.id == profile.id }
        sizes[profile.folderName] = nil
        saveMetadata()
    }

    func togglePin(_ profile: CursorProfile) {
        var p = profile
        p.isPinned.toggle()
        update(p)
    }

    func revealInFinder(_ profile: CursorProfile) {
        NSWorkspace.shared.activateFileViewerSelecting([directory(for: profile)])
    }

    // MARK: Launch / quit

    func launch(_ profile: CursorProfile, projectPath: String? = nil, memoryMB: Int? = nil, newWindow: Bool = false) {
        let options = CursorLauncher.LaunchOptions(
            memoryMB: memoryMB ?? profile.defaultMemoryMB,
            projectPath: projectPath ?? profile.defaultProjectPath,
            newWindow: newWindow
        )
        // Idempotent — also covers profiles adopted from disk that never went through createProfile.
        TitleBarColorizer.apply(profile: profile, profileDir: directory(for: profile))
        do {
            try CursorLauncher.launch(
                userDataDir: profile.isSystem ? nil : directory(for: profile),
                cursorPath: customCursorPath.isEmpty ? nil : customCursorPath,
                options: options
            )
            var p = profile
            p.lastLaunchedAt = Date()
            update(p)
            Task { [weak self] in
                try? await Task.sleep(for: .seconds(2))
                self?.refreshRunning()
            }
        } catch {
            lastError = error.localizedDescription
        }
    }

    func quit(_ profile: CursorProfile) {
        CursorLauncher.quit(pids: runningPIDs[profile.folderName] ?? [])
        Task { [weak self] in
            try? await Task.sleep(for: .seconds(1.5))
            self?.refreshRunning()
        }
    }

    func quitAll() {
        for profile in profiles where isRunning(profile) {
            CursorLauncher.quit(pids: runningPIDs[profile.folderName] ?? [])
        }
        Task { [weak self] in
            try? await Task.sleep(for: .seconds(1.5))
            self?.refreshRunning()
        }
    }

    /// Windows actually on screen for a profile (counted, not guessed).
    func windows(for profile: CursorProfile) -> Int {
        windowCounts[profile.folderName] ?? 0
    }

    // MARK: Background refreshers

    /// Refreshes "Running" status periodically, but only while a window is
    /// actually open and key — no reason to keep scanning every process on
    /// the Mac while the app sits unattended in the background.
    private func startPolling() {
        // Structured poll loop. The Task inherits MainActor isolation from
        // this method, so touching the store needs no annotations — and
        // crucially no explicit @MainActor on the closure, which would flip
        // checking to Sendable-capture rules and forbid even weak self.
        pollTask?.cancel()
        pollTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(15))
                guard let self, NSApp.isActive else { continue }
                self.refreshRunning()
            }
        }
    }

    private struct RunningSnapshot: Sendable {
        var pids: [String: [Int32]] = [:]
        var memory: [String: UInt64] = [:]
        var cpu: [String: Double] = [:]
        var windows: [String: Int] = [:]
    }

    func refreshRunning() {
        let items = profiles.map { ($0.folderName, directory(for: $0), $0.isSystem) }
        Task { [weak self] in
            let snapshot = await Task.detached(priority: .utility) {
                Self.computeRunningSnapshot(items: items)
            }.value
            self?.applyRunningSnapshot(snapshot)
        }
    }

    /// Pure computation: Sendable in, Sendable out, never touches the store.
    private static nonisolated func computeRunningSnapshot(
        items: [(folderName: String, dir: URL, isSystem: Bool)]
    ) -> RunningSnapshot {
        let processes = CursorLauncher.processList()
        let statsByPid = Dictionary(uniqueKeysWithValues: processes.map {
            ($0.pid, (rssKB: $0.rssKB, cpu: $0.cpuPercent))
        })
        var snapshot = RunningSnapshot()
        var allPIDs: [Int32] = []
        for item in items {
            let matched = item.isSystem
                ? CursorLauncher.systemProfilePIDs(in: processes)
                : CursorLauncher.pids(in: processes, profileDir: item.dir)
            snapshot.pids[item.folderName] = matched
            var rss: Int64 = 0
            var cpu = 0.0
            for pid in matched {
                rss += statsByPid[pid]?.rssKB ?? 0
                cpu += statsByPid[pid]?.cpu ?? 0
            }
            snapshot.memory[item.folderName] = UInt64(max(0, rss)) * 1024
            snapshot.cpu[item.folderName] = cpu
            allPIDs.append(contentsOf: matched)
        }
        let winsByPid = SystemStats.windowCounts(for: allPIDs)
        for (name, pids) in snapshot.pids {
            snapshot.windows[name] = pids.reduce(0) { $0 + (winsByPid[$1] ?? 0) }
        }
        return snapshot
    }

    @MainActor
    private func applyRunningSnapshot(_ snapshot: RunningSnapshot) {
        runningPIDs = snapshot.pids
        liveMemory = snapshot.memory
        liveCPU = snapshot.cpu
        windowCounts = snapshot.windows
        // host_statistics is microseconds-cheap; sample on main to
        // respect @MainActor isolation.
        if let sysCPU = cpuMonitor.usage() { systemCPU = sysCPU }
    }

    func refreshSizes() {
        let items = profiles.map { ($0.folderName, directory(for: $0)) }
        Task { [weak self] in
            let result = await Task.detached(priority: .utility) {
                var sizes: [String: Int64] = [:]
                for (name, dir) in items { sizes[name] = Self.directorySize(dir) }
                return sizes
            }.value
            if let self {
                for (name, size) in result { self.sizes[name] = size }
            }
        }
    }

    nonisolated private static func directorySize(_ url: URL) -> Int64 {
        let keys: Set<URLResourceKey> = [.totalFileAllocatedSizeKey, .isRegularFileKey]
        guard let enumerator = FileManager.default.enumerator(
            at: url, includingPropertiesForKeys: Array(keys),
            options: [], errorHandler: { _, _ in true }) else { return 0 }
        var total: Int64 = 0
        for case let file as URL in enumerator {
            guard let values = try? file.resourceValues(forKeys: keys),
                  values.isRegularFile == true else { continue }
            total += Int64(values.totalFileAllocatedSize ?? 0)
        }
        return total
    }
}
